File: mac/Marquee/API/Models/LibraryModels.swift
import Foundation

// Library status and the Sonarr/Radarr actions on a title (api-v1.md §4).

extension API {
    /// `library` on the title page and `GET …/status`.
    struct TitleLibraryInfo: Codable, Hashable, Sendable {
        /// Always shown as the title's status badge.
        let status: LibraryStatus
        /// Where ownership came from; nil when nothing has the title.
        let provider: LibraryProvider?
        /// The library owner's Radarr (movies) / Sonarr (TV) has a root folder and quality profile.
        let configured: Bool
        /// The "File details" card; non-nil for `owned` titles and for a show
        /// Sonarr has some episodes of.
        let file: FileDetails?
    }

    /// The "File details" card. Movie-only fields are nil for TV.
    struct FileDetails: Codable, Hashable, Sendable {
        /// "Location" (with Copy).
        let path: String?
        let sizeBytes: Int64
        /// The quality profile name, e.g. "Bluray-2160p".
        let quality: String?
        let resolutionTier: ResolutionTier?
        /// e.g. "3840x1600".
        let resolution: String?
        let videoCodec: String?
        let dynamicRange: String?
        let audioCodec: String?
        /// e.g. 7.1.
        let audioChannels: Double?
        let dateAdded: Date?
        let releaseGroup: String?
        let edition: String?
        /// "MKV"/"MP4" — from Plex or Jellyfin; the *arrs don't report it.
        let container: String?
        /// Whole-file bitrate in kbps, e.g. 58421. Media servers only.
        let bitrateKbps: Int?

        /// "29.1 GB".
        var sizeLabel: String { Format.bytes(sizeBytes) }

        /// The "Resolution" row: `resolutionTier ?? resolution`.
        var resolutionLabel: String? { resolutionTier?.rawValue ?? resolution.nonBlank }

        /// "Dynamic range": "Dolby Vision", "HDR10+", or as sent.
        var dynamicRangeLabel: String? { Quality.hdrLabel(dynamicRange) }

        /// "58.4 Mbps", or "820 kbps" below a megabit.
        var bitrateLabel: String? {
            guard let bitrateKbps, bitrateKbps > 0 else { return nil }
            if bitrateKbps < 1000 { return "\(bitrateKbps) kbps" }
            return String(format: "%.1f Mbps", Double(bitrateKbps) / 1000)
        }

        /// The "Audio" row: "TrueHD Atmos 7.1ch".
        var audioLabel: String? {
            guard let codec = audioCodec.nonBlank else { return nil }
            guard let audioChannels else { return codec }
            let channels = audioChannels.rounded() == audioChannels ? String(Int(audioChannels)) : String(audioChannels)
            return "\(codec) \(channels)ch"
        }
    }

    /// Radarr/Sonarr has the title (admin only).
    struct ArrTracking: Codable, Hashable, Sendable {
        /// The movie/series id inside Radarr/Sonarr.
        let arrId: Int
        let monitored: Bool
    }

    /// `viewer` on the title page: which actions to show under the title.
    struct TitleViewerState: Codable, Hashable, Sendable {
        let isAdmin: Bool
        /// The star (`favorites.set`).
        let favorited: Bool
        /// Your active request, if any.
        let requestStatus: RequestStatus?
        /// Show "Requested — waiting for approval" instead of Request.
        let alreadyRequested: Bool
        /// "Also requested by A, B" when non-empty and you haven't requested.
        let otherRequesters: [String]
        /// "Add to Radarr/Sonarr" (`titles.add`).
        let canAdd: Bool
        /// "Connect Radarr/Sonarr to add this title" (admin → Integrations).
        let needsArrSetup: Bool
        /// "Request" (`requests.create`).
        let canRequest: Bool
        /// "Wrong match? Fix ID" (`titles.relink`).
        let canRelink: Bool
        /// "Search now" and "Stop/Start monitoring" when non-nil.
        let arrTracking: ArrTracking?
        /// TV: at least one season can be requested (season picker, "Request
        /// more seasons"). nil from a server older than season requests,
        /// which only takes whole-series requests.
        let canRequestSeasons: Bool?
        /// The seasons of your pending request; nil when there's none or it's
        /// for the whole series.
        let requestedSeasons: [Int]?
        /// The 4K copy (0.37+): non-nil when the admin has a 4K Radarr (movies)
        /// / 4K Sonarr (TV). nil from an older server or without one.
        let fourK: FourKViewerState?
        /// "Report a problem" (0.38+): the title (or its 4K copy) is owned or
        /// downloading. nil from an older server, which takes no reports.
        let canReport: Bool?
        /// Your own open problem reports for this title (0.38+).
        let openReports: Int?
        /// The request blocklist (0.41+). When `.blocked`, `canRequest`,
        /// `canRequestSeasons` and `fourK.canRequest` are already false. nil
        /// from an older server, which can't block (no admin buttons then).
        let blocked: BlockState?
        /// Reviewers only (0.46+): Sonarr/Radarr hasn't found the approved
        /// request since then, the "Can't find" pill. nil otherwise, and
        /// from an older server.
        var notFoundSince: Date? = nil

        /// On the admin's blocklist: the member's "Requests are closed" pill,
        /// the admin's "Unblock requests".
        var block: TitleBlock? { blocked?.block }

        /// The admin's "Block requests" / "Unblock requests" is on offer.
        var offersBlocking: Bool { isAdmin && blocked != nil }

        /// Show the "Report a problem" button.
        var showsReportProblem: Bool { canReport == true }

        /// "Requested Seasons 1–3 — waiting for approval", or without the
        /// seasons for a whole-series request.
        var pendingRequestLine: String {
            if let label = API.seasonsLabel(requestedSeasons) {
                return "Requested \(label) — waiting for approval"
            }
            return "Requested — waiting for approval"
        }

        /// "Also requested by A, B", or nil when it shouldn't show.
        var otherRequestersLine: String? {
            guard !otherRequesters.isEmpty, !alreadyRequested else { return nil }
            return "Also requested by \(otherRequesters.joined(separator: ", "))"
        }
    }

    /// `viewer.fourK`: the title in the 4K Radarr/Sonarr (components/fourk-controls.tsx).
    struct FourKViewerState: Codable, Hashable, Sendable {
        /// How the 4K instance has the title, read live; `.untracked` when it doesn't.
        let status: LibraryStatus
        /// Your own 4K request (`pending`/`approved`); nil when none or declined.
        let requestStatus: RequestStatus?
        /// "Request in 4K" (`requests.create(…, is4k: true)`).
        let canRequest: Bool
        /// "Add to 4K Radarr/Sonarr" (`titles.add(…, is4k: true)`, admin).
        let canAdd: Bool

        /// The gold outline chip: "In 4K", "4K downloading", "4K missing",
        /// "4K coming soon"; nil when the 4K instance doesn't have it.
        var statusLabel: String? {
            switch status {
            case .owned: "In 4K"
            case .trackedDownloading: "4K downloading"
            case .trackedMonitored: "4K missing"
            case .comingSoon: "4K coming soon"
            case .untracked, .unknown: nil
            }
        }

        /// The "4K requested" chip: your 4K request is still pending.
        var isRequestPending: Bool { requestStatus == .pending }
    }

    /// `GET /titles/{type}/{tmdbId}/status`: just `library` + `viewer`, the
    /// cheap refresh after add / request / monitor / favorite.
    struct TitleStatus: Codable, Hashable, Sendable {
        let mediaType: MediaType
        let tmdbId: Int
        let library: TitleLibraryInfo
        let viewer: TitleViewerState

        var id: TitleID { TitleID(mediaType, tmdbId) }
    }

    /// `{"is4k": true}`: the body of "Request in 4K" and "Add to 4K Radarr/Sonarr".
    struct FourKBody: Encodable, Hashable, Sendable {
        var is4k = true
    }

    /// `PUT …/monitored` response.
    struct MonitoredResult: Codable, Hashable, Sendable {
        let ok: Bool
        let monitored: Bool
    }

    /// "Wrong match? Fix ID": the id to repoint the title to. The server
    /// checks `tmdbId`, then `imdbId`, then `tvdbId` (TV only).
    enum RelinkTarget: Encodable, Hashable, Sendable {
        case tmdb(Int)
        /// `tt0133093` or `0133093`.
        case imdb(String)
        case tvdb(Int)

        private enum CodingKeys: String, CodingKey {
            case tmdbId, imdbId, tvdbId
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)
            switch self {
            case let .tmdb(id): try container.encode(id, forKey: .tmdbId)
            case let .imdb(id): try container.encode(id, forKey: .imdbId)
            case let .tvdb(id): try container.encode(id, forKey: .tvdbId)
            }
        }
    }

    /// `POST …/relink` response: navigate to `newTmdbId` afterwards.
    struct RelinkResult: Codable, Hashable, Sendable {
        let ok: Bool
        let newTmdbId: Int
    }
}

File: windows/Marquee.Core/Models/LibraryModels.cs
using System.Globalization;

namespace Marquee.Core.Models;

// Library status and the Sonarr/Radarr actions on a title (api-v1.md
// section 4). Mirrors mac/Marquee/API/Models/LibraryModels.swift.

/// <summary><c>library</c> on the title page and <c>GET …/status</c>.</summary>
public sealed record TitleLibraryInfo
{
    /// <summary>Always shown as the title's status badge.</summary>
    public required LibraryStatus Status { get; init; }

    /// <summary>Where ownership came from; null when nothing has the title.</summary>
    public LibraryProvider? Provider { get; init; }

    /// <summary>The library owner's Radarr (movies) / Sonarr (TV) has a root folder and quality profile.</summary>
    public required bool Configured { get; init; }

    /// <summary>The "File details" card; non-null only for <c>owned</c> titles.</summary>
    public FileDetails? File { get; init; }
}

/// <summary>
/// The "File details" card. Movie-only fields are null for TV, and which
/// fields are filled depends on what owns the title (see the doc's table):
/// where both an *arr and a media server know a field, the *arr's value
/// wins, since it is authoritative about the release it fetched.
/// </summary>
public sealed record FileDetails
{
    /// <summary>"Location" (with Copy).</summary>
    public string? Path { get; init; }

    public required long SizeBytes { get; init; }

    /// <summary>The quality profile name, e.g. "Bluray-2160p".</summary>
    public string? Quality { get; init; }

    public ResolutionTier? ResolutionTier { get; init; }

    /// <summary>e.g. "3840x1600".</summary>
    public string? Resolution { get; init; }

    public string? VideoCodec { get; init; }

    /// <summary><c>"DV" | "HDR10" | "HDR10Plus" | "HLG" | "PQ" | "SDR"</c>; null means "not known", not "SDR".</summary>
    public string? DynamicRange { get; init; }

    public string? AudioCodec { get; init; }

    /// <summary>e.g. 7.1.</summary>
    public double? AudioChannels { get; init; }

    public DateTimeOffset? DateAdded { get; init; }
    public string? ReleaseGroup { get; init; }
    public string? Edition { get; init; }

    /// <summary>"MKV"/"MP4", from Plex or Jellyfin; the *arrs don't report it (older servers omit the key).</summary>
    public string? Container { get; init; }

    /// <summary>Whole-file bitrate in kbps, e.g. 58421. Media servers only (older servers omit the key).</summary>
    public int? BitrateKbps { get; init; }

    /// <summary>"29.1 GB".</summary>
    public string SizeLabel => FormatBytes(SizeBytes);

    /// <summary>The "Resolution" row: <c>resolutionTier ?? resolution</c>.</summary>
    public string? ResolutionLabel => ResolutionTier?.Value ?? Resolution.NonBlank();

    /// <summary>"Dynamic range": "Dolby Vision", "HDR10+", or as sent.</summary>
    public string? DynamicRangeLabel => HdrLabel(DynamicRange);

    /// <summary>"58.4 Mbps", or "820 kbps" below a megabit.</summary>
    public string? BitrateLabel
    {
        get
        {
            if (BitrateKbps is not { } kbps || kbps <= 0)
            {
                return null;
            }
            return kbps < 1000
                ? $"{kbps.ToString(CultureInfo.InvariantCulture)} kbps"
                : $"{(kbps / 1000.0).ToString("0.0", CultureInfo.InvariantCulture)} Mbps";
        }
    }

    /// <summary>The "Audio" row: "TrueHD Atmos 7.1ch".</summary>
    public string? AudioLabel
    {
        get
        {
            if (AudioCodec.NonBlank() is not { } codec)
            {
                return null;
            }
            if (AudioChannels is not { } channels)
            {
                return codec;
            }
            // "2ch", not "2.0ch": the website drops a whole number's fraction.
            var count = Math.Round(channels) == channels
                ? ((int)channels).ToString(CultureInfo.InvariantCulture)
                : channels.ToString(CultureInfo.InvariantCulture);
            return $"{codec} {count}ch";
        }
    }

    /// <summary>Port of lib/format.ts's <c>formatBytes</c>: one decimal, 1024-based units.</summary>
    private static string FormatBytes(long bytes)
    {
        if (bytes <= 0)
        {
            return "0 B";
        }
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        var exponent = Math.Min((int)Math.Floor(Math.Log(bytes) / Math.Log(1024)), units.Length - 1);
        var value = bytes / Math.Pow(1024, exponent);
        return $"{value.ToString("0.0", CultureInfo.InvariantCulture)} {units[exponent]}";
    }

    /// <summary>Port of lib/quality.ts's <c>hdrLabel</c>.</summary>
    private static string? HdrLabel(string? dynamicRange)
    {
        if (string.IsNullOrEmpty(dynamicRange))
        {
            return null;
        }
        if (string.Equals(dynamicRange, "dv", StringComparison.OrdinalIgnoreCase))
        {
            return "Dolby Vision";
        }
        if (string.Equals(dynamicRange, "hdr10plus", StringComparison.OrdinalIgnoreCase))
        {
            return "HDR10+";
        }
        return dynamicRange;
    }
}

/// <summary>Radarr/Sonarr has the title (admin only).</summary>
public sealed record ArrTracking
{
    /// <summary>The movie/series id inside Radarr/Sonarr.</summary>
    public required int ArrId { get; init; }

    public required bool Monitored { get; init; }
}

/// <summary><c>viewer</c> on the title page: which actions to show under the title.</summary>
public sealed record TitleViewerState
{
    public required bool IsAdmin { get; init; }

    /// <summary>The star (<c>PUT</c>/<c>DELETE /favorites/{type}/{tmdbId}</c>).</summary>
    public required bool Favorited { get; init; }

    /// <summary>Your active request, if any.</summary>
    public RequestStatus? RequestStatus { get; init; }

    /// <summary>Show the website's "Requested, waiting for approval" state instead of the Request button.
    public required bool AlreadyRequested { get; init; }

    /// <summary>"Also requested by A, B" when non-empty and you haven't requested.</summary>
    public required IReadOnlyList<string> OtherRequesters { get; init; }

    /// <summary>"Add to Radarr/Sonarr" (<c>POST …/add</c>).</summary>
    public required bool CanAdd { get; init; }

    /// <summary>"Connect Radarr/Sonarr to add this title" (admin: a link to Integrations).</summary>
    public required bool NeedsArrSetup { get; init; }

    /// <summary>"Request" (<c>POST …/request</c>).</summary>
    public required bool CanRequest { get; init; }

    /// <summary>"Wrong match? Fix ID" (<c>POST …/relink</c>).</summary>
    public required bool CanRelink { get; init; }

    /// <summary>"Search now" and "Stop/Start monitoring" when non-null.</summary>
    public ArrTracking? ArrTracking { get; init; }

    /// <summary>
    /// TV: at least one season can be requested (season picker, "Request more
    /// seasons"). Null from a server older than season requests, which only
    /// takes whole-series requests.
    /// </summary>
    public bool? CanRequestSeasons { get; init; }

    /// <summary>The seasons of your pending request; null when there's none or it's for the whole series.</summary>
    public IReadOnlyList<int>? RequestedSeasons { get; init; }

    /// <summary>
    /// The 4K copy (0.37+), when the admin has set up a 4K Radarr (movies) /
    /// 4K Sonarr (TV); null otherwise, and from an older server.
    /// </summary>
    public FourKViewerState? FourK { get; init; }

    /// <summary>
    /// "Report a problem" shows (0.38+): the title, or its 4K copy, is owned
    /// or downloading. An older server leaves it out, which hides it.
    /// </summary>
    public bool CanReport { get; init; }

    /// <summary>Your own open problem reports for this title (0.38+; 0 when left out): "Problem reported" instead of the button.</summary>
    public int OpenReports { get; init; }

    /// <summary>"Requested Seasons 1–3, waiting for approval", or without the seasons for a whole-series request.</summary>
    public string PendingRequestLine =>
        SeasonLabels.SeasonsLabel(RequestedSeasons) is { } label
            ? $"Requested {label}, waiting for approval"
            : "Requested, waiting for approval";

    /// <summary>"Also requested by A, B", or null when it shouldn't show.</summary>
    public string? OtherRequestersLine =>
        OtherRequesters.Count == 0 || AlreadyRequested ? null : $"Also requested by {string.Join(", ", OtherRequesters)}";
}

/// <summary>
/// <c>viewer.fourK</c> (0.37+): the title in the 4K Radarr/Sonarr, read live
/// and independent of <c>library.status</c>. components/fourk-controls.tsx.
/// </summary>
public sealed record FourKViewerState
{
    /// <summary>How the 4K instance has the title; <see cref="LibraryStatus.Untracked"/> when it doesn't.</summary>
    public required LibraryStatus Status { get; init; }

    /// <summary>Your own 4K request (pending/approved); null when none or declined.</summary>
    public RequestStatus? RequestStatus { get; init; }

    /// <summary>"Request in 4K" (<c>POST …/request</c> with <c>{"is4k": true}</c>).</summary>
    public required bool CanRequest { get; init; }

    /// <summary>"Add to 4K Radarr/Sonarr" (admin; <c>POST …/add</c> with <c>{"is4k": true}</c>).</summary>
    public required bool CanAdd { get; init; }

    /// <summary>The gold outline chip: "In 4K", "4K downloading", "4K missing", "4K coming soon"; null when untracked (or unknown).</summary>
    public string? StatusLabel
    {
        get
        {
            if (Status == LibraryStatus.Owned) return "In 4K";
            if (Status == LibraryStatus.TrackedDownloading) return "4K downloading";
            if (Status == LibraryStatus.TrackedMonitored) return "4K missing";
            if (Status == LibraryStatus.ComingSoon) return "4K coming soon";
            return null;
        }
    }

    /// <summary>The "4K requested" chip, while your 4K request is pending.</summary>
    public bool IsRequested => RequestStatus == Models.RequestStatus.Pending;
}

/// <summary><c>POST …/request</c> and <c>POST …/add</c> body for the 4K copy (0.37+).</summary>
public sealed record FourKBody(bool Is4k);

/// <summary>
/// <c>GET /titles/{type}/{tmdbId}/status</c>: just <c>library</c> +
/// <c>viewer</c>, the cheap refresh after add / request / monitor / favorite.
/// </summary>
public sealed record TitleStatus
{
    public required MediaType MediaType { get; init; }
    public required int TmdbId { get; init; }
    public required TitleLibraryInfo Library { get; init; }
    public required TitleViewerState Viewer { get; init; }

    public TitleId Id => new(MediaType, TmdbId);
}

/// <summary><c>PUT …/monitored</c> body.</summary>
public sealed record SetMonitoredRequest(bool Monitored);

/// <summary><c>PUT …/monitored</c> response.</summary>
public sealed record MonitoredResult
{
    public required bool Ok { get; init; }
    public required bool Monitored { get; init; }
}

/// <summary>
/// "Wrong match? Fix ID": the id to repoint the title to. Exactly one of the
/// three is set, and only that key goes on the wire; the server checks
/// <c>tmdbId</c>, then <c>imdbId</c>, then <c>tvdbId</c> (TV only).
/// </summary>
public sealed record RelinkTarget
{
    private RelinkTarget()
    {
    }

    public int? TmdbId { get; init; }

    /// <summary><c>tt0133093</c> or <c>0133093</c>.</summary>
    public string? ImdbId { get; init; }

    public int? TvdbId { get; init; }

    public static RelinkTarget Tmdb(int id) => new() { TmdbId = id };
    public static RelinkTarget Imdb(string id) => new() { ImdbId = id };
    public static RelinkTarget Tvdb(int id) => new() { TvdbId = id };
}

/// <summary><c>POST …/relink</c> response: navigate to <see cref="NewTmdbId"/> afterwards.</summary>
public sealed record RelinkResult
{
    public required bool Ok { get; init; }
    public required int NewTmdbId { get; init; }
}
